src/lib.rs

```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
     }
 }
 
+#[cfg(feature = "serde")]
+pub fn versioning_from_string<'de, D>(deserializer: D) -> Result<Versioning, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+
+    Versioning::new(&s)
+        .ok_or_else(|| Err(Error::IllegalVersioning(s)))
+        .map_err(D::Error::custom)
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+/// A version requirement expression, like `^1.4.163`. Equivalent to a `VersionReq` for
+/// a `Versioning` instead of a semver `Version`.
+pub struct VersioningReq {
+    /// The version requirement operation, or `None` if the operation is `*`.
+    op: Op,
+    /// The version requirement expression
+    version: Option<Versioning>,
+}
+
+impl VersioningReq {
+    /// Create a new `VersioningReq` from an operation and a version.
+    ///
+    /// # Arguments
+    ///
+    /// * `op` - The operation to use for the requirement.
+    /// * `version` - The version to use for the requirement. If `None`, the version is empty,
+    ///              which happens when the operation is `*`.
+    pub fn new(op: Op, version: Option<Versioning>) -> Self {
+        Self { op, version }
+    }
+
+    /// Return the operation of the requirement.
+    pub fn op(&self) -> Op {
+        self.op
+    }
```

Comment:
Why not just make the struct fields `pub` and leave these methods out?

Reply:
Sounds good!